### PR TITLE
Add ability to assert metadata properties on input dataset parameters

### PR DESCRIPTION
**Pull request by bernt-matthias**
revive https://github.com/galaxyproject/galaxy/pull/3159 from @jmchilton  (anno 2016 .. just a little bit more than crazy 10K PRs/issues in between  :))

Changed the name for the validator .. `dataset` prefix might be good for comparability.

Not sure if one could / should generalize / reuse `dataset_metadata_in_range` (which considers only numbers at the moment).

ping @bgruening 

Side note: I would find this super userful to check for column names, but `tabular` does not set them .. still downstream tools could and et voila it can be  used :). 

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

File: lib/galaxy/tools/parameters/validation.py [hunk 1]
```diff
@@ -949,6 +986,7 @@ def validate(self, value, trans=None):
     in_range=InRangeValidator,
     length=LengthValidator,
     metadata=MetadataValidator,
+    dataset_metadata_equal=MetadataEqualsValidator,
```
Comment (jmchilton) [inline lib/galaxy/tools/parameters/validation.py:989]: Maybe ``metadata_equal`` for consistency with the above type?

What about stronger typing. I feel like I've been heading this way with tool stuff... if ``value`` is used treat it as a string and use a ``value_json`` attribute to ``json.load`` the value for a typed comparison.

I would make these two changes but you write more tools - if you prefer the way you have it here that is totally fine with me.

File: test/functional/tools/metadata_check_eq.xml [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+            <param name="input" value="1.tabular" ftype="tabular" />
+        </test>
+        <test>
+            <param name="input" value="1.csv" ftype="csv" />
+            <output name="output">
+                <assert_contents>
+                    <has_text text="sucess"/>
+                    <has_text text="columns 12"/>
+                    <!-- unfortunately it seems impossible to access metadata and ,-join in cheetah -->
+                    <has_text text="column_names __ob____sq__Transaction_date__sq__, __sq__Product__sq__, __sq__Price__sq__, __sq__Payment_Type__sq__, __sq__Name__sq__, __sq__City__sq__, __sq__State__sq__, __sq__Country__sq__, __sq__Account_Created__sq__, __sq__Last_Login__sq__, __sq__Latitude__sq__, __sq__Longitude__sq____cb__"/>
```
Comment (jmchilton) [inline test/functional/tools/metadata_check_eq.xml]: This feels over fit - how about just ``<has_text text="Latitude" />`` here and ``<not_has_text text="Latitude" />`` below.

File: lib/galaxy/tools/parameters/validation.py (hunk 1 above)
Comment (bernt-matthias) [reply lib/galaxy/tools/parameters/validation.py:989]: > Maybe `metadata_equal` for consistency with the above type?

Hmm. We also have `dataset_metadata_...` maybe we can simplify/unify all this by adding a few aliases and name all of them `metadata_...`?

> What about stronger typing. I feel like I've been heading this way with tool stuff... if `value` is used treat it as a string and use a `value_json` attribute to `json.load` the value for a typed comparison.

`value_json` is a great idea. I vaguely remembered that I have seen something like this but was not sure. I will add it.

> I would make these two changes but you write more tools - if you prefer the way you have it here that is totally fine with me.

Thanks for the feedback. Give me a moment :)

Comment (bernt-matthias) [reply lib/galaxy/tools/parameters/validation.py:989]: While thinking about this I would suggest to stick with `dataset_metadata_equal`: All the user (ie tool developer) facing validators are prefixed with `dataset_metadata_`. The `metadata` validator is unlikely to be used by tool developers (its automatically added to each data parameter). So maybe this is an acceptable level of inconsistency...

File: lib/galaxy/tool_util/linters/inputs.py [hunk 3]
```diff
@@ -392,6 +397,18 @@ def lint_inputs(tool_xml, lint_ctx):
                     f"Parameter [{param_name}]: '{vtype}' validators need to define the 'min' or 'max' attribute(s)",
                     node=validator,
                 )
+            if vtype in ["dataset_metadata_equal"]:
+                if not ("value" in validator.attrib or "value_json" in validator.attrib) or "metadata_name" not in validator.attrib:
```
Comment (bernt-matthias) [inline lib/galaxy/tool_util/linters/inputs.py]:
```suggestion
                if (
                    not ("value" in validator.attrib or "value_json" in validator.attrib)
                    or "metadata_name" not in validator.attrib
                ):
```

**Comment by mvdbeek** [comment]
This needs a pass of black.

**Comment by bernt-matthias** [comment]
Good to go?

File: lib/galaxy/tools/parameters/validation.py [hunk 4]
```diff
@@ -559,6 +560,42 @@ def validate(self, value, trans=None):
             super().validate(isinstance(value, model.DatasetInstance) and not missing, value_to_show=missing)
 
 
+class MetadataEqualsValidator(Validator):
```
Comment (nsoranzo) [inline lib/galaxy/tools/parameters/validation.py:563]: I'm late, but maybe `MetadataEqualValidator` instead of `MetadataEqualsValidator` for consistency?

Comment (bernt-matthias) [reply lib/galaxy/tools/parameters/validation.py:563]: Not to late https://github.com/galaxyproject/galaxy/pull/16489 :)

Thanks!

**Pull request merged**